src/gennylib/include/gennylib/topology.hpp

```diff
@@ -0,0 +1,211 @@
+
+    // Called before/after/between visiting mongoses.
+    virtual void onBeforeMongoses(const ShardedDescription&) {}
+    virtual void onAfterMongoses(const ShardedDescription&) {}
+    virtual void onBetweenMongoses(const ShardedDescription&) {}
+
+    virtual ~TopologyVisitor() = 0;
+};
+
+// Be careful changing the traversal order of the cluster; visitors may depend on it.
+struct TopologyDescription {
+    virtual void accept(TopologyVisitor&) = 0;
+    virtual ~TopologyDescription() = 0;
+};
+
+struct MongodDescription : public TopologyDescription {
+    std::string mongodUri;
+
+    void accept(TopologyVisitor& v) { v.onMongod(*this); }
+};
+
+struct MongosDescription : public TopologyDescription {
+    std::string mongosUri;
+
+    void accept(TopologyVisitor& v) { v.onMongos(*this); }
+};
+
+struct ReplSetDescription : public TopologyDescription {
+    std::string primaryUri;
+    bool configsvr = false;
+    std::vector<MongodDescription> nodes;
+
+    void accept(TopologyVisitor& v) { 
+        v.onBeforeReplSet(*this); 
+        for (int i = 0; i < nodes.size() - 1; i++) {
+            nodes[i].accept(v);
+            v.onBetweenMongods(*this);
+        }
+        nodes[nodes.size() - 1].accept(v);
+        v.onAfterReplSet(*this); 
+    }
+};
+
+struct ShardedDescription : public TopologyDescription {
+    ReplSetDescription configsvr;
+    std::vector<ReplSetDescription> shards;
+    std::vector<MongosDescription> mongoses;
+
+    void accept(TopologyVisitor& v) { 
+        v.onBeforeSharded(*this); 
+        configsvr.accept(v);
+
+        v.onBeforeShards(*this); 
+        for (int i = 0; i < shards.size() - 1; i++) {
+            shards[i].accept(v);
+            v.onBetweenShards(*this);
+        }
+        shards[shards.size() - 1].accept(v);
+        v.onAfterShards(*this); 
+
+        v.onBeforeMongoses(*this);
+        for (int i = 0; i < mongoses.size() - 1; i++) {
+            mongoses[i].accept(v);
+            v.onBetweenMongoses(*this);
+        }
+        mongoses[mongoses.size() - 1].accept(v);
+        v.onAfterMongoses(*this);
+
+        v.onAfterSharded(*this); 
+    }
+};
+
+/**
+ * Represents the topology of a MongoDB cluster.
+ */
+class Topology {
+public:
+    Topology(mongocxx::client& client) : _factory{client.uri().to_string()} {
+        MongoConnection connection(client.uri().to_string());
+        update(connection);
+    }
+
+    Topology(DBConnection& connection) : _factory{connection.uri()} {
+        update(connection);
+    }
+
+    /**
+     * Traverse the cluster, using the visitor to act on it.
+     */
+    void accept(TopologyVisitor& v) { 
+        if (_topologyDesc) {
+            v.onBeforeTopology(*_topologyDesc);
+            _topologyDesc->accept(v); 
+            v.onAfterTopology(*_topologyDesc);
+        }
+    }
+
+    /**
+     * Update the Topology's view of the cluster.
+     */
+    void update(DBConnection& connection);
```

Comment:
It should be safe to pass-by-reference since `update` doesn't store `DBConnection` anywhere and the enclosing scope outlives the `update` function? My concern with pass-by-value is that we don't want to create a new connection to the server, we want to reuse the existing one. And I think we don't want to have the `DBConnection` use `shard_ptr` to store the `mongocxx::client`?